### PR TITLE
Grab bag of fixes for ROCm 5

### PR DESCRIPTION
**Pull request by e-kayrakli**
Resolves https://github.com/chapel-lang/chapel/issues/23735
Supersedes https://github.com/chapel-lang/chapel/pull/23897

- Separates `rocm-version.h` header to be able to set `ROCM_VERSION_MAJOR` correctly.
- Adjusts PATH when
  - compiling `gpu-amd-reduce.cc` during runtime build
  - compiling `.c` files through `chpl`
  - This is to resolve https://github.com/chapel-lang/chapel/issues/23735
  - (We need to bundle lld and have system LLVM build with that _or_ use ROCm's bundled LLVM as the ideal solution here)
- Removes `test/gpu/native/rocmOnly`: the test in there wasn't testing anything that a user would do. Moreover, it caused some issues with ROCm 5, that I wasn't too interested in resolving.
- Fixes `p2p` test to expect for `XGMI` in the output of `rocm-smi` as the connection type
- Throws in `--fast` in `coral` even in correctness setting. I can't tell why, but without it, this code runs extremely slowly. The setup we had didn't finish in 5 minutes. Note that this test will also change soon via https://github.com/chapel-lang/chapel/pull/23913

Test:
- [x] amd with rocm 5.4
- [x] amd with rocm 4.4
- [x] nvidia

### CODE REVIEW

**Comment by e-kayrakli** [comment]
@jabraham17 -- this PR has a bunch of minor stuff, but the most significant change is resolving your https://github.com/chapel-lang/chapel/issues/23735.

File: runtime/src/gpu/common/rocm-version.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+#if __has_include(<rocm-core/rocm_version.h>)  // 5.x wants this
+#include <rocm-core/rocm_version.h>
+#elif __has_include(<rocm/rocm_version.h>)  // 4.x wants this
+#include <rocm/rocm_version.h>
+#elif __has_include(<rocm_version.h>)  // Deprecated. 5.x used to want this
+#include <rocm_version.h>
+#elif !defined(ROCM_VERSION_MAJOR)
+#define ROCM_VERSION_MAJOR 4 // this is the safe bet
+#endif
```
Comment (jabraham17) [inline runtime/src/gpu/common/rocm-version.h:35]: If I understand this right, we are searching headers for `ROCM_VERSION_MAJOR`, and if its not set defaulting to 4? So should `#elif !defined(ROCM_VERSION_MAJOR)` still be apart of the elif chain?

File: test/gpu/native/studies/bandwidth/p2p.chpl [hunk 2]
```diff
@@ -112,8 +112,11 @@ proc checkNumAccessEnabled_rocm(valFromChpl) {
   var sub = spawn(["rocm-smi", "--showtopotype"], stdout=pipeStyle.pipe);
   var line : string;
   var count = 0;
-  while sub.stdout.readLine(line) do
-    count += line.count("PCIE");
+  while sub.stdout.readLine(line) {
+    // these two shold be mutually exclusive
```
Comment (jabraham17) [inline test/gpu/native/studies/bandwidth/p2p.chpl:116]: nit, spelling on the comment

File: test/gpu/native/studies/coral/coral.compopts [hunk 3]
```diff
@@ -1 +1,4 @@
--M modules
+# for some reason, this test runs extermely slowly on ROCm 5 without --fast.
```
Comment (jabraham17) [inline test/gpu/native/studies/coral/coral.compopts:1]: nit, spelling on the comment

File: compiler/llvm/clangUtil.cpp [hunk 4]
```diff
@@ -4955,6 +4969,11 @@ void makeBinaryLLVM(void) {
       }
     }
 
+    if (usingGpuLocaleModel() &&
+        getGpuCodegenType() == GpuCodegenType::GPU_CG_AMD_HIP) {
+      setenv("PATH", curPath.c_str(), /*override*/ 1);
+    }
+
```
Comment (jabraham17) [inline compiler/llvm/clangUtil.cpp:4976]: Why is this necessary? You do the same above when setting `curPath`?

File: compiler/llvm/clangUtil.cpp [hunk 5]
```diff
@@ -4936,17 +4936,31 @@ void makeBinaryLLVM(void) {
       cargs += clangInfo->clangCCArgs[i];
     }
 
+    // if we are using rocm, we'll have to adjust the path so that clang can
+    // find rocm's lld. We ideally want to include lld in our bundled llvm
+    std::string curPath = "";
+
```
Comment (jabraham17) [inline compiler/llvm/clangUtil.cpp:4942]: See other comment, but if `curPath` is only set and used from within a single `else if` block, can you move this variable declaration and comment there?

**Review by jabraham17 — APPROVED** [review]
Looks good!

File: runtime/src/gpu/common/rocm-version.h (hunk 1 above)
Comment (e-kayrakli) [reply runtime/src/gpu/common/rocm-version.h:35]: That's right. But I am not following... which part of what I have feels wrong? Would you expect this to be just `else`? If that's the case, what I have feels safer in case:

1. this file was included multiple times, or,
2. `ROCM_VERSION_MAJOR` was set via other means like a `-D` or something.

(2) is not a big worry obviously, but (1) is.

File: compiler/llvm/clangUtil.cpp (hunk 4 above)
Comment (e-kayrakli) [reply compiler/llvm/clangUtil.cpp:4976]: This part is setting `PATH` back to its original value. See the difference in arguments (`adjPath` vs `curPath`)

File: runtime/src/gpu/common/rocm-version.h (hunk 1 above)
Comment (jabraham17) [reply runtime/src/gpu/common/rocm-version.h:35]: I would have expected something like
```c++
#if __has_include(<rocm-core/rocm_version.h>)  // 5.x wants this
#include <rocm-core/rocm_version.h>
#elif __has_include(<rocm/rocm_version.h>)  // 4.x wants this
#include <rocm/rocm_version.h>
#elif __has_include(<rocm_version.h>)  // Deprecated. 5.x used to want this
#include <rocm_version.h>
#endif

#if !defined(ROCM_VERSION_MAJOR)
#define ROCM_VERSION_MAJOR 4 // this is the safe bet
#endif
```

File: compiler/llvm/clangUtil.cpp (hunk 5 above)
Comment (e-kayrakli) [reply compiler/llvm/clangUtil.cpp:4942]: Answered in https://github.com/chapel-lang/chapel/pull/24018#discussion_r1416419036

File: runtime/src/gpu/common/rocm-version.h (hunk 1 above)
Comment (e-kayrakli) [reply runtime/src/gpu/common/rocm-version.h:35]: Oh, I see. OK, yeah I like that better.

Comment (e-kayrakli) [reply runtime/src/gpu/common/rocm-version.h]: https://github.com/chapel-lang/chapel/pull/24018/commits/79fbc96d07e83e0e7a0f28e795cdf5d4b1a69880
